command/init.go

```diff
@@ -771,6 +783,53 @@ func (c *InitCommand) getProviders(config *configs.Config, state *states.State,
 			))
 		}
 
+		// Legacy providers required by state which could not be installed are
+		// probably in-house providers. If the user has completed the necessary
+		// steps to make their custom provider available for installation, then
+		// there should be a provider with the same type selected after the
+		// installation process completed.
+		//
+		// If we detect this specific situation, we can confidently suggest
+		// that the next step is to run the state replace-provider command to
+		// update state. We build a map of provider replacements here to ensure
+		// that we're as concise as possible with the diagnostic.
+		stateReplaceProviders := make(map[addrs.Provider]addrs.Provider)
+		for provider, fetchErr := range stateLegacyProviderErrors {
+			var sameType []addrs.Provider
+			for p := range selected {
+				if p.Type == provider.Type {
+					sameType = append(sameType, p)
+				}
+			}
+			if len(sameType) == 1 {
+				stateReplaceProviders[provider] = sameType[0]
+			} else {
+				diags = diags.Append(tfdiags.Sourceless(
+					tfdiags.Error,
+					"Failed to install provider",
+					fmt.Sprintf("Error while installing %s: %s", provider.ForDisplay(), fetchErr),
+				))
+			}
+		}
+		if len(stateReplaceProviders) > 0 {
+			var detail strings.Builder
+			command := "command"
+			if len(stateReplaceProviders) > 1 {
+				command = "commands"
+			}
+
+			fmt.Fprintf(&detail, "Found unresolvable legacy provider references in state. It looks like these refer to in-house providers. You can update the resources in state with the following %s:\n", command)
+			for legacy, replacement := range stateReplaceProviders {
+				fmt.Fprintf(&detail, "\n    terraform state replace-provider \\\n      %s \\\n      %s", legacy, replacement)
```

Comment:
I have a feeling this backslash-escaping of newlines won't work in the Windows command interpreter :confounded: . Although it'll be ugly to read, I think it'd be safer to put this all on one line so that everyone can just copy-paste it into whatever shell they are using. :disappointed:

Reply:
Oh sorry I just noticed in your last comment that you did this to avoid the automatic line wrapping. :roll_eyes: 

I thought we had a rule in the line wrapping that it would not try to wrap a line that starts with leading whitespace, but I guess I imagined it. With that said, I think it could be reasonable to add such an exception, if that would make this easier.

Reply:
Thanks for the prompt! I hadn't considered trying to adjust how we format diagnostics to allow longer command lines. I made an attempt in 3547f9e and updated the other commits accordingly.